### ISSUE TITLE
Loading large grid views takes long

### ISSUE THREAD

**Issue by constantinpape** [mobie/mobie-viewer-fiji#256]
Loading the large grid views in the zebrafish dataset takes very long from local internet.
Is there any way to do this in a more lazy fashion?
I.e. get all the shapes and compute the new viewer transform and switch to it.
Only then start to lazily load the sources.

Is there maybe some other blocking calculation going on that could be sped up?
(I can check next week from EMBL intranet how much faster it gets with a proper internet connection)

**Comment by tischi** [comment]
I completely agree with the slow speed!
I will look into making this more lazy...

**Comment by tischi** [comment]
I am not sure what to do, if the user zooms out we will anyway have to load all of them and it would be slow.

I also had a look and implementing additional laziness is a bit annoying because, as you anticipated, the GridSourceTransformer needs the sources in order to transform them.... 

But I admit right now nothing happening for a long time is a bit disturbing. 

Maybe a progress report in the BDV window would help?

**Comment by constantinpape** [comment]
> I am not sure what to do, if the user zooms out we will anyway have to load all of them and it would be slow.

But couldn't that happen lazily as usual in BDV?

> I also had a look and implementing additional laziness is a bit annoying because, as you anticipated, the GridSourceTransformer needs the sources in order to transform them....
> 
> But I admit right now nothing happening for a long time is a bit disturbing.
> 
> Maybe a progress report in the BDV window would help?

In theory it should work to compute the grid transforms based on the extents after transformation though (and for this you wouldn't need to load the data), or am I missing something?

Anyway, I agree that this is not really highest priority right now and adding some progress bar or some indication that the viewer is busy would help a lot already.

**Comment by tischi** [comment]
> In theory it should work to compute the grid transforms based on the extents after transformation though (and for this you wouldn't need to load the data), or am I missing something?

I need the `SourceAndConverter` objects, because these are the objects to be transformed, and, normally without introducing additional non-standard laziness, these object can only be instantiated by first loading the `SpimData`; not the whole data, of course, but whatever this code loads initially (I am not sure how much this loads, maybe only the xml, maybe more...):

```
InputStream stream = FileAndUrlUtils.getInputStream(path);
SpimData spimData = (new CustomXmlIoSpimData()).loadFromStream(stream, path);
return spimData;
```

...and then there is code in bdv-playground that creates a `SourceAndConverter` object from a `SpimData` object.
In fact, since the `SourceAndConverter` has a defined pixel type (e.g. uint16),  I guess this means that some part of the code must have accessed the S3 store, because the xml does not contain information about the pixel type (even we did not show the image yet in BDV).

I have one idea (inspired by our conversation to @bogovicj): For the gridView I will try to parallelize the code that fetches the `SpimData` and creates the `SourceAndConverter` for all the images in the grid, maybe that is faster than doing it sequentially. 

Note: in principle, once shown in BDV everything should be lazy and parallel (as many threads as CPUs); but I also noticed BDV sometimes getting a bit stuck (not very unresponsive to, e.g., mouse drag) when loading several images in parallel from S3. I don't know whether there still is something that could be improved in bdv-core. We would have to show this to @tpietzsch, I guess.

**Comment by constantinpape** [comment]
Ok, this indeed sounds fairly complicated...
I think if we add some indication that the viewer is busy it should work for now, but we should leave this one open as future work.

**Comment by tischi** [comment]
Yeah! (cc @bogovicj)
I managed to open and transform the sources in parallel.
Significant speed up 🥳 
(My crappy internet at home)

```
Fetched 24 image source(s) in 21833 ms, using 1 threads.
Transformed 24 image source(s) in 5093 ms, using 1 threads.
Fetched 24 image source(s) in 22347 ms, using 1 threads.
Transformed 24 image source(s) in 5296 ms, using 1 threads.
```

```
Fetched 24 image source(s) in 4935 ms, using 4 thread(s).
Transformed 24 image source(s) in 1538 ms, using 4 thread(s).
Fetched 24 image source(s) in 5423 ms, using 4 thread(s).
Transformed 24 image source(s) in 2396 ms, using 4 thread(s).
```

```
Fetched 24 image source(s) in 3325 ms, using 8 thread(s).
Transformed 24 image source(s) in 975 ms, using 8 thread(s).
Fetched 24 image source(s) in 2990 ms, using 8 thread(s).
Transformed 24 image source(s) in 1146 ms, using 8 thread(s).
```

```
Fetched 24 image source(s) in 3442 ms, using 16 thread(s).
Transformed 24 image source(s) in 653 ms, using 16 thread(s).
Fetched 24 image source(s) in 2405 ms, using 16 thread(s).
Transformed 24 image source(s) in 1092 ms, using 16 thread(s).
```

My comp has only 4 cores. I think I will hard-code 8 threads for now.

**Comment by constantinpape** [comment]
This has been much improved now, closing the issue.
